File: C-Sharp/sort/SelectionSort.cs
/*
C# Program to Perform a Selection Sort

Selection sort is an algorithm of sorting an array where it loop from the start 
of the loop, and check through other elements to find the minimum value.
After the end of the first iteration, the minimum value is swapped with the current element.
The iteration then continues from the 2nd element and so on.
*/
using System;

class SelectionSort
{
    static void Main(string[] args)
    {
        //Take array size input
        Console.Write("Enter array size: ");
        int array_size = Convert.ToInt32(Console.ReadLine());
        
        int[] array = new int[array_size]; 
        
        //Take array elements input
        for(int i=0; i<array_size; i++){
            Console.Write("Element {0} : ", i);
            array[i] = Convert.ToInt32(Console.ReadLine());
        }
        
        //Print array before Selection sort
        Console.Write("\nThe Array Before Selection Sort is: ");
        for(int i=0; i<array_size; i++)
        {
            Console.Write(array[i]+ " ");
        }
        
        int MinIndex = 0;//current index of the current minimum wwe are looking for
        int temp = 0; //temparory variable for swapping the value of the current min index
 
        //MainIndex is the pointer for each lowest we need to find
        for(int MainIndex=0; MainIndex<array_size-1; MainIndex++)
        {
            //MinIndex is now the current spot in the main list
            MinIndex = MainIndex;
            
            //remaining goes threw the remainder of the list finding the index of the next lowest value
            for(int RemainingIndex=MainIndex+1; RemainingIndex<array_size; RemainingIndex++)
            {
                //if the Remaining element is the less then the MinIndex we move the index to that spot
                //will swap values after going through the entire list
                if (array[RemainingIndex] < array[MinIndex])
                {
                    //Update minimum amount index
                    MinIndex = RemainingIndex;
                }
            }
 
            //We now have the index of the lowest number in the remaining list. Swap it withthe current i element in the list  
            temp = array[MinIndex];
            array[MinIndex] = array[MainIndex];
            array[MainIndex] = temp;
        }
 
        Console.Write("\nThe Array After Selection Sort is: ");
        for (int i=0; i<array_size; i++)
        {
            Console.Write(array[i]+ " ");
        }
    }
}
/*
Input:

Enter array size: 5
Element 0 : 67
Element 1 : 34
Element 2 : 56
Element 3 : 12
Element 4 : 73

Output:
The Array Before Selection Sort is: 67 34 56 12 73 
The Array After Selection Sort is: 12 34 56 67 73

*/

/*
Time Complexity: O(n2)
Space Complexity: O(1)
*/

File: PHP/sort/SelectionSort.php
<?php
/*Selection Sort is a type of sorting algorithm which saves the index of the
first element in every iteration and checks whether there is an element smaller than it,
then it saves the index of that particular element and swap it with the first element
of each iteration*/

    // function to implement Selection Sort in PHP
  
function SelectionSort(&$arr) {
    for ($i = 0; $i < count($arr) - 1; $i++)
    {   
    //the position of the first element is choosen as the minimum index
   //at each iterations of outer loop
        
        $min_index = $i;
        
        //then the below loop will run to check whether there is a
        //element smaller than the element at $min_index
        //if so it will swap the element at index j
        
        //with the element at min_index
    for ($j = $i + 1; $j < count($arr) - 1; $j++)
    {
    if ($arr[$j] < $arr[$min_index])
    {
    $min_index = $j;
    }
    }
        //swapping is done on the basis of the index found
        //which is stored in index min_index
        
        //and then swapped
    $temp = $arr[$min_index];
    $arr[$min_index] = $arr[$i];
    $arr[$i] = $temp;
        }
       }
       
$arr = array(-1, 2, -8, 854, -19, 0, 25000);
echo "Array before sorting:"."\n";
for($i=0;$i<count($arr);$i++)
{
    echo $arr[$i]." ";
}
SelectionSort($arr);
echo "\nArray after sorting:"."\n";
for($i=0;$i<count($arr);$i++)
{
    echo $arr[$i]." ";
}

//time complexity :O(n^2)
//space complexity:O(1)


/* Sample Output
Array before sorting:
-1 2 -8 854 -19 0 25000
Array after sorting:
-19 -8 -1 0 2 854 25000
*/
?>
